[merged] Add cluster subcommand

mbarnes opened the pull request:
This adds a `cluster` subcommand to integrate with Commissaire, as discussed in https://github.com/projectatomic/atomic/issues/329.

One thing to note about `atomic cluster` is it does **not** require root privilege; it's just talking over HTTP.  I tried to solve that cleanly in followup commits by delaying the root user check until after the command line is parsed and then adding a way for subcommands to opt out of requiring the root user.  This has the added benefit of letting non-root users at least get `--help` output for any `atomic` command.

mbarnes commented:
CC @ashcrow: Extra pair of eyeballs wouldn't hurt, if you're interested.

ashcrow commented:
@mbarnes looks good to me!

rhatdan commented on bash/atomic:
```diff
@@ -651,12 +651,109 @@ _atomic_migrate_import() {
 
 }
 
+_atomic_cluster_no_opts() {
```
Can you create this in Atomic/cluster.py?

mbarnes replied:
@rhatdan I can, but are you sure?  Nothing else under `Atomic/` appears to be serving bash completion.  Just wanted to double check first.

rhatdan replied:
Never mind, I though this was part of the atomic command not the bash completions.

rhatdan commented on atomic:
```diff
@@ -119,6 +116,15 @@ if __name__ == '__main__':
+    # atomic cluster (if available)
+    try:
+        import commctl.cli
+        clusterp = subparser.add_parser(
```
Not crazy on this, since I think we should just require it.  Having it be optional makes us shipping the man page seem bad.  Remove the try block.

mbarnes replied:
No problem.  For the record, though, I was following the example of `atomic host` which is also written to be optional.

rhatdan replied:
Yes that is true, but that is more of a case where we are differentiating from normal OS and Atomic Host.

This cluster command will be available on the normal OS correct?  IE RHEL as well as RHEL atomic host?

mbarnes replied:
Correct, it will be on RHEL and RHEL Atomic Host.

jlebon replied:
I don't follow this, why should we require it? Atomic on its own is useful without commissaire (e.g. I like using atomic on my workstation, but I'm not planning to join a cluster anytime soon).

rhatdan commented:
How about this patch for running atomic as non root.

https://github.com/projectatomic/atomic/pull/355

mbarnes commented:
> How about this patch for running atomic as non root.
> #355

Looks like it already got merged, but I'm happy with it too.  I'll rebase this branch.

rhatdan commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,65 @@
+
+# SUBCOMMANDS
+**get cluster** CLUSTER_NAME
+  Report status of the specified cluster.
```
Shouldn't this be just get not get cluster

atomic cluster get CLUSTER_NAME

mbarnes replied:
Hmm... I agree that particular case reads better.  But there's some other types of things you can "get" (or "create" or "delete") besides cluster objects and I'm worried it would make parsing tricky -- or even ambiguous if we add more types of things to "get" in the future.

An easier solution is to rename the `atomic cluster` part, though the best I can think of is `atomic multihost`.

@ashcrow: Any thoughts here?

ashcrow replied:
There are other options to ```commctl```'s```cluster``` command (```get```, ```create```, ```delete```). However, seeing ```cluster get cluster $NAME``` does seem odd.

A few ideas:

* Maybe it would make better sense to use a different direct command than ```cluster```. Maybe ```manage```?
* Maybe we can flatten out the commands we add so that they are at a higher level like:

```
atomic cluster get $NAME
atomic host delete $NAME
atomic status # maybe this one is a special case where we say atomic manager status
...
```

rhatdan replied:
Perhaps we should setup a bluejeans to discuss this.  I really do not understand what Commissaire does, so tough for me to give opinion.

mbarnes replied:
>   Maybe we can flatten out the commands we add so that they are at a higher level like:

> atomic cluster get $NAME
> atomic host delete $NAME
> atomic status # maybe this one is a special case where we say atomic manager status

That's going to trample on the `atomic` namespace too much I think.  `atomic host`, for example, is already a thing (wraps `rpm-ostree`).  I'm on board with your first suggestion though, but picking a good name is hard.

One thing that maybe wasn't stated explicitly enough in https://github.com/projectatomic/atomic/issues/329 was that we already have a standalone command-line tool for this named `commctl` and it shares the same command syntax (and code) with what we're proposing for `atomic cluster`.

IOW, `atomic cluster get cluster` == `commctl get cluster`.

But it's not clear to me (or even @ashcrow I think) whether `commctl` is meant to be kept if/when this stuff lands in `atomic`.

rhatdan commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,65 @@
+  Report status of the specified cluster.
+
+**get host** IP_ADDRESS
+  Report status of the host with the specified IP address.
```
DITTO

rhatdan commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,65 @@
+  Report status of the host with the specified IP address.
+
+**get restart** CLUSTER_NAME
+  Report status of a restart operation on the specified cluster.
```
DITTO

rhatdan commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,65 @@
+  Report status of a restart operation on the specified cluster.
+
+**get upgrade** CLUSTER_NAME
+  Report status of an upgrade operation on the specified cluster.
```
DITTO

rhatdan commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,65 @@
+  Report status of an upgrade operation on the specified cluster.
+
+**create cluster** CLUSTER_NAME
+  Register a new cluster with the specified name.
```
DITTO

rhatdan commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,65 @@
+  Register a new cluster with the specified name.
+
+**create host** [**-c**/**--cluster** CLUSTER_NAME] IP_ADDRESS SSH_PRIV_KEY
+  Register a new host with the specified IP address.  SSH_PRIV_KEY is a file path to the host's private SSH key (e.g. id_rsa), used to drive operations on the host.  Use **--cluster** to add the host to an existing cluster.
```
DITTO

mbarnes commented:
@rhatdan, @ashcrow: I know we agreed on a syntax when we chatted, but the `atomic host` thing still isn't sitting right with me and I think I know why now.

Currently `atomic host` commands all act on the *local* host through `rpm-ostree`, whereas these new commands all act on *remote* hosts through Commissaire.

I think conflating local vs remote actions into the same `atomic host` prefix will be confusing to users.

@ashcrow and I were tossing around more ideas and we hit upon making it more explicit that a subset of `atomic` commands apply to *remote* hosts.

What do you think about these?
```
atomic remote cluster get|create|delete|list
atomic remote cluster reboot|upgrade start|status
atomic remote host get|create|delete|list
```
(@ashcrow, I took some liberties here vs what's in `commctl`)

cgwalters commented:
Seems sane to me.  On a semi-related topic, Cockpit ships `remotectl` which I always thought was clever, but AFAIK it was never used for anything more than the cockpit cert.

rhatdan commented:
Is the cluster remote  Or just the host?
Aren't I just joining the cluster.

Not crazy about 4 commands.  
atomic remote host delete foobar?

mbarnes commented:
> Is the cluster remote Or just the host?
> Aren't I just joining the cluster.

You *can* add hosts to a cluster manually, but I believe the intent is for a cluster node to add itself on boot via cloud-init (which just uses Python libraries) and reserve the `commctl`/`atomic whatever` command for admins on a machine *outside* the cluster. (@ashcrow, right?)

> Not crazy about 4 commands.

> atomic remote host delete foobar?

Yeah, it's a little wordy.  Would hyphenating help?
```
atomic remote-host delete foobar
```

rhatdan commented:
How about 
atomic rhost delete foobar
atomic cluster get foobar

ashcrow commented:
rhost sounds good to me.

mbarnes commented:
I'm good with that too.  It will require a bit of rework on the `commctl` side, so I'll do that first.

rhatdan commented:
Any update?

mbarnes commented:
Not quite yet.  Got side tracked, but I'm back on this now.

mbarnes commented:
@rhatdan Before I start on docs / bash completion updates here, can you take a look at https://github.com/projectatomic/commctl/pull/12?  That shows the new syntax for `commctl`.

`atomic` would be identical except `rhost` instead of `host`.

rhatdan commented:
That looks good to me, except for the `commctl passhash`, Are you going to make a atomic equivalent?

mbarnes commented:
> That looks good to me, except for the commctl passhash, Are you going to make a atomic equivalent?

That one seems like a poor fit for `atomic` to me.  It's just a helper script.  I'm inclined to omit it for now and just leave it in `commctl`.  If `commctl` ends up going away then maybe reevaluate.

Sound alright?

rhatdan commented:
Fine with me.  Your the boss.

ashcrow commented:
I'm fine with keeping ```passhash``` in ```commctl``` only.

mbarnes commented:
:arrow_up: Complete rewrite works with latest `commctl`

 (which the Jenkins environment appears to be missing)

mbarnes commented:
Rebased and updated bash completion + manpages for a new Commissaire operation coming soon.  (https://github.com/projectatomic/commissaire/pull/135)

I don't know how to correct for the Jenkins failure.  The `commctl` package this feature needs is either outdated or missing in the test environment.

rhatdan commented:
@jlebon How do we get the latest version of commictl into the test environment?

jlebon commented:
Shouldn't the import and cli parsers be conditional on commissaire being present? E.g. like we did for the recent system containers feature and ostree.

mbarnes commented:
> Shouldn't the import and cli parsers be conditional on commissaire being present? E.g. like we did for the recent system containers feature and ostree.

I had it conditional at first and some discussion ensued.  See outdated patch review comments above from Apr 23 (can't seem to link to them directly...)

jlebon commented:
Yup, commented there as well, but you're right, the [link](https://github.com/projectatomic/atomic/pull/354#discussion_r65389095) to it doesn't actually bring you to the discussion. Reposting here to make it easier:

> I don't follow this, why should we require it? Atomic on its own is useful without commissaire (e.g. I like using atomic on my workstation, but I'm not planning to join a cluster anytime soon).

rhatdan commented:
I am fine with going back to making it optional, just not crazy about man pages being available but functionality not.  But we set the president of ostree, so I guess we can follow it here.

cgwalters commented:
Talking about this in terms of Jenkins is mostly a red herring - there's two types of dependencies:

 - Build time
 - Runtime

(And each type can be either hard or soft/optional)

The current internal Jenkins PR tester conflates these, but the rpmdistro-gitoverlay builder does not, it uses RPM semantics.  (On the other hand it doesn't yet run per-project `make check`)

One major pain point specific to this project is the way we extract the version number from the python code is that it turns all hard runtime dependencies into build-time too.

Philosophy aside it does seem like having the dep be soft/optional at runtime makes sense.

mbarnes commented:
:arrow_up:  Jenkins is happy again.

We could revisit the optional aspect once `commctl` is available where we need it.

jlebon commented on docs/atomic-cluster.1.md:
```diff
@@ -0,0 +1,52 @@
+# SYNOPSIS
+**atomic cluster [OPTIONS] SUBCOMMAND**
+
+**atomic cluster** will issue commands to a Commissaire server over HTTP.
```
Maybe a small line here to say that it will only be available if Commissaire is installed?

mbarnes replied:
I'm not sure where manpages get packaged for AH these days but my thought was if we could temporarily delete these two pages in a spec file or whatever, that seems cleaner to me than a disclaimer.  Failing that, yeah I can add a line.

rhatdan replied:
They are packaged up in the RHELtools container

jlebon commented on docs/atomic-rhost.1.md:
```diff
@@ -0,0 +1,34 @@
+# SYNOPSIS
+**atomic rhost [OPTIONS] SUBCOMMAND**
+
+**atomic rhost** will issue commands to a Commissaire server over HTTP.
```
And here

rhatdan commented:
What is the latest on this pull request?

ashcrow commented:
I'm :+1: on @jlebon's suggestions. @mbarnes can you rebase?

mbarnes commented:
:arrow_up: I added the man page disclaimers @jlebon suggested, rebased on current `master` and squashed my previous `fixup!` comments.  Should be good to go.

jlebon commented:
:+1:
@rh-atomic-bot r+ 2581030
